File: ios/SecureCore/Sources/SecureCore/DocumentService.swift
import Foundation
import Security

/// Errors specific to document service operations.
public enum DocumentServiceError: Error, Equatable {
    case encryptionFailed(String)
    case decryptionFailed(String)
    case randomGenerationFailed
    case metadataNotFound(docId: String)
}

/// Orchestration service that assembles all SecureCore components to provide
/// high-level document import, decrypt, list, and delete operations.
///
/// Guarantees:
/// - DEKs are zeroized immediately after use.
/// - On import failure, partial state (file / metadata) is cleaned up.
/// - Atomic writes via the underlying `DocumentStoreProtocol`.
public final class DocumentService {
    private static let dekLength = 32

    private let secureCoreLib: SecureCoreLibProtocol
    private let keyManager: KeyManagerProtocol
    private let documentStore: DocumentStoreProtocol
    private let metadataRepository: MetadataRepositoryProtocol

    public init(
        secureCoreLib: SecureCoreLibProtocol,
        keyManager: KeyManagerProtocol,
        documentStore: DocumentStoreProtocol,
        metadataRepository: MetadataRepositoryProtocol
    ) {
        self.secureCoreLib = secureCoreLib
        self.keyManager = keyManager
        self.documentStore = documentStore
        self.metadataRepository = metadataRepository
    }

    // MARK: - Public API

    /// Imports a document: generates a DEK, encrypts, stores the blob and metadata.
    /// Returns the generated document ID.
    public func importDocument(data: Data, filename: String, mimeType: String) async throws
        -> String
    {
        let docId = UUID().uuidString

        // Generate DEK
        var dek = try generateRandomBytes(count: Self.dekLength)
        defer { dek.resetBytes(in: dek.startIndex..<dek.endIndex) }

        // Encrypt
        let blob: Data
        do {
            blob = try secureCoreLib.encryptBytes(data, dek: dek)
        } catch {
            throw DocumentServiceError.encryptionFailed(error.localizedDescription)
        }

        // Wrap DEK
        let wrapsJson: Data
        do {
            wrapsJson = try keyManager.wrapDek(dek)
        } catch {
            throw error
        }

        // Store encrypted blob
        do {
            try documentStore.writeDocument(docId: docId, data: blob)
        } catch {
            throw error
        }

        // Save metadata
        let record = DocumentRecord(
            docId: docId,
            filename: filename,
            mimeType: mimeType,
            createdAt: Int64(Date().timeIntervalSince1970 * 1000),
            plaintextSize: Int64(data.count),
            ciphertextSize: Int64(blob.count),
            wrapsJson: String(data: wrapsJson, encoding: .utf8) ?? ""
        )

        do {
            try metadataRepository.save(record)
        } catch {
            // Cleanup: remove the stored blob since metadata save failed
            _ = try? documentStore.deleteDocument(docId: docId)
            throw error
        }

        return docId
    }

    /// Decrypts a document and returns the plaintext bytes.
    public func decryptDocument(docId: String) async throws -> Data {
        let record = try findRecordOrThrow(docId: docId)
        let blob = try documentStore.readDocument(docId: docId)

        guard let wrapsData = record.wrapsJson.data(using: .utf8) else {
            throw DocumentServiceError.decryptionFailed("Invalid wrapsJson encoding")
        }

        var dek = try keyManager.unwrapDek(wrapsData)
        defer { dek.resetBytes(in: dek.startIndex..<dek.endIndex) }

        do {
            return try secureCoreLib.decryptBytes(blob, dek: dek)
        } catch {
            throw DocumentServiceError.decryptionFailed(error.localizedDescription)
        }
    }

    /// Decrypts a document and writes the plaintext to a temporary file.
    /// The caller is responsible for deleting the file when done.
    public func decryptDocumentToTempFile(docId: String, tempDir: URL) async throws -> URL {
        let plaintext = try await decryptDocument(docId: docId)
        let record = try findRecordOrThrow(docId: docId)

        let ext = fileExtension(for: record.mimeType)
        let filename = UUID().uuidString + ext
        let fileURL = tempDir.appendingPathComponent(filename)

        try plaintext.write(to: fileURL, options: .atomic)
        return fileURL
    }

    /// Lists all document metadata records.
    public func listDocuments() async throws -> [DocumentRecord] {
        try metadataRepository.list()
    }

    /// Deletes a document's encrypted file and metadata.
    public func deleteDocument(docId: String) async throws {
        _ = try documentStore.deleteDocument(docId: docId)
        try metadataRepository.delete(docId: docId)
    }

    // MARK: - Private

    private func findRecordOrThrow(docId: String) throws -> DocumentRecord {
        guard let record = try metadataRepository.find(docId: docId) else {
            throw DocumentServiceError.metadataNotFound(docId: docId)
        }
        return record
    }

    private func generateRandomBytes(count: Int) throws -> Data {
        var bytes = [UInt8](repeating: 0, count: count)
        let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
        guard status == errSecSuccess else {
            throw DocumentServiceError.randomGenerationFailed
        }
        return Data(bytes)
    }

    private func fileExtension(for mimeType: String?) -> String {
        switch mimeType?.lowercased() {
        case "application/pdf": return ".pdf"
        case "image/jpeg": return ".jpg"
        case "image/png": return ".png"
        case "text/plain": return ".txt"
        default: return ".bin"
        }
    }
}

File: android/secure-core-android/src/main/kotlin/com/securecore/DocumentService.kt
package com.securecore

import com.securecore.keymanager.KeyManager
import com.securecore.metadata.DocumentEntity
import com.securecore.metadata.MetadataRepository
import com.securecore.store.DocumentStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.InputStream
import java.security.SecureRandom
import java.util.UUID

/**
 * Central orchestrator for document encryption workflows.
 *
 * Coordinates [CryptoEngine], [KeyManager], [DocumentStore], and
 * [MetadataRepository] to provide atomic import/decrypt/delete operations.
 *
 * DEKs are zeroized immediately after use.
 */
class DocumentService(
    private val cryptoEngine: CryptoEngine,
    private val keyManager: KeyManager,
    private val documentStore: DocumentStore,
    private val metadataRepository: MetadataRepository
) {

    suspend fun importDocument(
        inputStream: InputStream,
        filename: String,
        mimeType: String
    ): Result<String> = withContext(Dispatchers.IO) {
        val docId = UUID.randomUUID().toString()
        val dek = ByteArray(32)
        try {
            SecureRandom().nextBytes(dek)

            val plaintext = inputStream.readBytes()
            val plaintextSize = plaintext.size.toLong()

            val blob = when (val result = cryptoEngine.encryptBytes(plaintext, dek)) {
                is SecureCoreResult.Success -> result.value
                is SecureCoreResult.Error -> return@withContext Result.failure(result.error)
            }

            val wrappedDek = keyManager.wrapDek(dek)
            dek.fill(0)

            documentStore.writeDocument(docId, blob)

            val entity = DocumentEntity(
                docId = docId,
                filename = filename,
                mimeType = mimeType,
                createdAt = System.currentTimeMillis(),
                plaintextSize = plaintextSize,
                ciphertextSize = blob.size.toLong(),
                contentHash = null,
                wrappedDek = wrappedDek,
                recoveryWrap = null,
                wrapAlgorithm = "AES-GCM-KEYSTORE"
            )

            when (val saveResult = metadataRepository.save(entity)) {
                is SecureCoreResult.Success -> {}
                is SecureCoreResult.Error -> {
                    documentStore.deleteDocument(docId)
                    return@withContext Result.failure(saveResult.error)
                }
            }

            Result.success(docId)
        } catch (e: Exception) {
            dek.fill(0)
            documentStore.deleteDocument(docId)
            Result.failure(e)
        }
    }

    suspend fun decryptDocument(docId: String): Result<ByteArray> = withContext(Dispatchers.IO) {
        val dek = ByteArray(0) // placeholder
        try {
            val entity = when (val result = metadataRepository.get(docId)) {
                is SecureCoreResult.Success -> result.value
                    ?: return@withContext Result.failure(
                        SecureCoreError.InvalidParameter("Document not found: $docId")
                    )
                is SecureCoreResult.Error -> return@withContext Result.failure(result.error)
            }

            val unwrappedDek = keyManager.unwrapDek(entity.wrappedDek)
            try {
                val blob = documentStore.readDocument(docId)

                val plaintext = when (val result = cryptoEngine.decryptBytes(blob, unwrappedDek)) {
                    is SecureCoreResult.Success -> result.value
                    is SecureCoreResult.Error -> return@withContext Result.failure(result.error)
                }

                Result.success(plaintext)
            } finally {
                unwrappedDek.fill(0)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun decryptDocumentToTempFile(
        docId: String,
        tempDir: File
    ): Result<File> = withContext(Dispatchers.IO) {
        when (val result = decryptDocument(docId)) {
            is Result -> {
                val plaintext = result.getOrElse { return@withContext Result.failure(it) }
                try {
                    tempDir.mkdirs()
                    val file = File(tempDir, "$docId.preview")
                    file.writeBytes(plaintext)
                    file.deleteOnExit()
                    Result.success(file)
                } catch (e: Exception) {
                    Result.failure(e)
                }
            }
        }
    }

    suspend fun listDocuments(): Result<List<DocumentMetadata>> = withContext(Dispatchers.IO) {
        when (val result = metadataRepository.list()) {
            is SecureCoreResult.Success -> Result.success(
                result.value.map { entity ->
                    DocumentMetadata(
                        docId = entity.docId,
                        filename = entity.filename,
                        mimeType = entity.mimeType,
                        createdAt = entity.createdAt,
                        plaintextSize = entity.plaintextSize,
                        ciphertextSize = entity.ciphertextSize
                    )
                }
            )
            is SecureCoreResult.Error -> Result.failure(result.error)
        }
    }

    suspend fun deleteDocument(docId: String): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            documentStore.deleteDocument(docId)
            when (val result = metadataRepository.delete(docId)) {
                is SecureCoreResult.Success -> Result.success(Unit)
                is SecureCoreResult.Error -> Result.failure(result.error)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
